### PR TITLE
drivers: pwm_nrf5_sw: Fix the driver and make it support nRF9x SoCs

### LINKED ISSUE #15323: blink_led sample does not work on most of the nRF boards
The issue was originally reported on slack (#general):
```
Aaron Tsui 9:15 AM, March 22nd
BTW, As the PWM_CLK lowest is 125khz(16M/128) and COUNTERTOP max is 0x7fff on nrf52, my understanding the nrf52 HW PWM longest period is about 262ms, cannot reach period as 1000ms. right?
@gaute this will cause failure on sample/blink_led, which need a 1000ms period.
```
Indeed, when trying to run this sample on any board with an nRF52 series SoC, which by default uses the HW PWM driver, one hits the following error:
```
PWM demo app-blink LED
pwm pin set fails
[00:00:00.008,544] <err> pwm_nrfx: Prescaler for period_cycles 16000000 not found.

[00:00:00.008,544] <err> pwm_nrfx: Incompatible period 16000000
```
Same happens for boards with nRF91 series SoCs.

For nRF52 the solution could be to use the nRF SW PWM driver, for nRF91 the driver is unavailable.

### PR DESCRIPTION
**Pull request by anangl**
This PR makes the blink_led sample working (again) on boards with nRF52 series SoCs. The SW PWM driver is selected for them instead of the default HW one, as the latter cannot handle periods longer than ~262 ms (this is a hardware limitation).
It also fixes a couple of issues (details below) found in the pwm_nrf5_sw driver and makes it support nRF9x SoCs as well, so that the blink_led sample can be run on boards equipped with those.

**drivers: pwm_nrf5_sw: Use dynamically adjusted timer prescaler**

Adjust the timer prescaler used by the pwm_nrf5_sw driver according to
the requested period length, so that it does not need to be specified
at the build time. Remove the incorrect "scaling" that was done to the
requested pulse and period cycles to fit them into 16 bits.

**samples: blink_led: Use nRF SW PWM driver for nRF52 boards**

The nrfx Hardware PWM driver cannot be used in this sample
since the driver does not allow to set a period longer than ~262 ms.

**drivers: pwm_nrf5_sw: Correct the use of GPIOTE channels and TIMER**

GPIOTE base channel number taken from DeviceTree was not used
consistenly, hence the driver could properly work only when this
number had the value of 0.
TIMER was not stopped during the reconfiguration of the channel
and this could lead to the inversion of the PWM polarity.
The config structure was removed on the occasion as it unnecessarily
complicated the code of this currently single-instance only driver.

**drivers: pwm_nrf5_sw: Use nrfx HALs instead of direct register accesses**

Now instead of accessing peripheral registers directly, the driver
uses the nrfx HAL functions. This makes the code easier to maintain
and also gives the possibility to use pins from GPIO port P1 on
nRF52840 (absolute numbers 32-47) as the GPIO HAL provides the needed
translation.

**drivers: pwm_nrf5_sw: Refactor the code**

Move all channel initialization code that does not need to be executed
on every pin reconfiguration (like initialization of the PPI channels)
to the driver initialization function. Simplify the routines that check
the requested period and find a suitable channel for driving the pin.

**drivers: pwm_nrf5_sw: Make the driver support nRF9x series SoCs as well**

Now the driver uses DPPI or PPI depending on the presence of one of
these interfaces in a given SoC, hence it can be used on nRF9x SoCs
as well. Remove the '5' from its name then, and make the blink_led
sampe use it for the nrf9160_pca10090 board, thus making the sample
working properly on this board.

Fixes #15323.

### CODE REVIEW

**Comment by anangl** [comment]
Fixed checkpatch and codeowners issues.

**Comment by ioannisg** [comment]
Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK

**Comment by anangl** [comment]
> Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK

I am not aware of any. The bug was reported on slack:
```
Aaron Tsui 9:15 AM, March 22nd
BTW, As the PWM_CLK lowest is 125khz(16M/128) and COUNTERTOP max is 0x7fff on nrf52, my understanding the nrf52 HW PWM longest period is about 262ms, cannot reach period as 1000ms. right?
@gaute this will cause failure on sample/blink_led, which need a 1000ms period.
```

**Comment by anangl** [comment]
cc @overheat

**Comment by overheat** [comment]
@anangl @ioannisg I didn't raise an issue specific for this topic. But I think this is related to #14749, which is verifying samples. Since this is a basic sample bug, I think it's better to be solved before the LTS version released.

**Comment by anangl** [comment]
Added one line that I somehow lost during splitting the changes into several logical commits. Sorry about that.

**Comment by anangl** [comment]
@ioannisg 
> > Is there a filed bug for this PR? Otherwise, it shouldn't go to 1.14 - we are close to release, AFAIK
> 
> I am not aware of any. The bug was reported on slack:

[..] 

> Sure. Could you file an issue in GH with "bug" and 1.14" that relates to the slack report?

Created #15323.

**Comment by anangl** [comment]
Removed a no longer needed overlay file that was changing the `clock-prescaler` property eliminated by this PR.

File: CODEOWNERS [hunk 1]
```diff
@@ -218,7 +218,6 @@
 /lib/posix/                               @pfalcon
 /kernel/device.c                          @andrewboie @andyross @nashif
 /kernel/idle.c                            @andrewboie @andyross @nashif
-/samples/basic/servo_motor/*microbit*     @jhe
```
Comment (MaureenHelm) [inline CODEOWNERS]: Looks like a stray change that doesn't belong in this commit

File: drivers/pwm/pwm_nrf5_sw.c [hunk 2]
```diff
@@ -208,12 +230,11 @@ static int pwm_nrf5_sw_pin_set(struct device *dev, u32_t pwm,
...
 {
-	struct pwm_config *config;
-
-	config = (struct pwm_config *)dev->config->config_info;
-
-	/* HF timer frequency is derived from 16MHz source with a prescaler */
-	*cycles = 16000000UL / BIT(config->prescaler);
+	/* Since this function might be removed (see issue #6958), the maximum
+	 * supported frequency (16 MHz) is always returned here, and dynamically
+	 * adjusted timer prescalers are used in the pin set function.
```
Comment (MaureenHelm) [inline drivers/pwm/pwm_nrf5_sw.c]: Not sure I understand this logic. Might the application want to know the actual frequency after scaling, perhaps to know the current timing range/resolution?

File: dts/arm/nordic/nrf52810.dtsi [hunk 3]
```diff
@@ -200,7 +200,7 @@
 &sw_pwm {
 	timer-instance = <2>;
 	channel-count = <3>;
-	clock-prescaler = <0>;
 	ppi-base = <14>;
 	gpiote-base = <0>;
+	status = "ok";
```
Comment (MaureenHelm) [inline dts/arm/nordic/nrf52810.dtsi]: Better to do this at the board level instead.

File: samples/basic/blink_led/boards/nrf52810_pca10040.conf [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# The nrfx Hardware PWM driver cannot be used in this sample since the driver
+# does not allow to set a period longer than ~262 ms.
+CONFIG_PWM_NRFX=n
```
Comment (MaureenHelm) [inline samples/basic/blink_led/boards/nrf52810_pca10040.conf]: Can you add to this comment that you're disabling the HW PWM driver in order to enable the SW PWM driver?

File: drivers/pwm/pwm_nrf5_sw.c [hunk 5]
```diff
@@ -14,41 +14,34 @@
...
-#define PWM_0_MAP_SIZE DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT
+#define PWM_MAP_SIZE  DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT
 
 /* Nordic TIMER peripherals allow prescalers 0-9. */
 #define MAX_TIMER_PRESCALER  9
 /* Nordic TIMERs can be 16- or 32-bit wide. */
 #define MAX_TIMER_VALUE  (GET_TIMER_CAPABILITY(MAX_SIZE) == 32 ? UINT32_MAX \
 							       : UINT16_MAX)
 
-struct pwm_config {
```
Comment (MaureenHelm) [inline drivers/pwm/pwm_nrf5_sw.c]: Can the hardware support multiple instances? If so, this is a step in the wrong direction. Why not fix the driver to use the device tree data consistently?

**Review by MaureenHelm — DISMISSED** [review]

File: samples/basic/blink_led/boards/nrf52810_pca10040.conf (hunk 4 above)
Comment (anangl) [reply samples/basic/blink_led/boards/nrf52810_pca10040.conf]: Done.

File: CODEOWNERS (hunk 1 above)
Comment (anangl) [reply CODEOWNERS]: I have split the commit into two, to make the removal of the "clock-prescaler" property and the above entry from CODEOWNERS more visible and clear.

File: dts/arm/nordic/nrf52810.dtsi (hunk 3 above)
Comment (anangl) [reply dts/arm/nordic/nrf52810.dtsi]: Actually, I'm not convinced it would be better. The activation of the nRF SW PWM driver is done through Kconfig option, so this enabling here leads only to generation of `DT_` macros, which is harmless. And it's more convenient to enable the node here than in all boards that uses this SoC. Am I missing some important consequence of doing it this way?

File: drivers/pwm/pwm_nrf5_sw.c (hunk 5 above)
Comment (anangl) [reply drivers/pwm/pwm_nrf5_sw.c]: The driver actually combines several peripherals to achieve the PWM functionality. But yes, it can be made multi-instance (though, due to the checks of capabilities of the selected TIMER, this will probably require more macro magic). I was not sure if it would be desirable and didn't want to do this in this PR. The removal of this structure was to clean up the current state (the driver was single-instance from its origin).

File: drivers/pwm/pwm_nrf5_sw.c (hunk 2 above)
Comment (anangl) [reply drivers/pwm/pwm_nrf5_sw.c]: We use the same approach in [pwm_nrfx](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/pwm/pwm_nrfx.c#L236). It is much more convenient for the users to have the prescaler computed dynamically (see the discussion in #5469 and the referred #6958) instead of having to set it at the build time.

**Comment by henrikbrixandersen** [comment]
Do these change also address the issue raised in #8628 (Fix polarity issue when using multiple channels) in the SW PWM driver?

**Comment by anangl** [comment]
@henrikbrixandersen 
> Do these change also address the issue raised in #8628 (Fix polarity issue when using multiple channels) in the SW PWM driver?

Yes, they do. By [stopping the timer](https://github.com/zephyrproject-rtos/zephyr/pull/15287/files#diff-bc88d4b3519c4c50a750394fa1c95146R172). But I think I'll try to use the solution with PPI channel group, as proposed by @cvinayak in https://github.com/zephyrproject-rtos/zephyr/pull/8628#discussion_r210176762.

**Comment by anangl** [comment]
Applied a couple of minor corrections.

**Comment by anangl** [comment]
Rebased on master.

**Comment by anangl** [comment]
Applied adjustments needed after rebasing.

**Comment by anangl** [comment]
Rebased on master to resolve the conflict in `samples/basic/servo_motor/src/main.c`.

**Comment by carlescufi** [comment]
@MaureenHelm can you please take a fresh look at this one?

**Comment by carlescufi** [comment]
@cvinayak @kl-cruz @nordic-krch can you please take a look?

File: drivers/pwm/pwm_nrf_sw.c [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+#include <nrf_gpiote.h>
+#if defined(PPI_PRESENT)
+#include <nrfx_ppi.h>
+#else
+#include <nrfx_dppi.h>
+#endif
+
+#define LOG_LEVEL CONFIG_PWM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pwm_nrf_sw);
```
Comment (nordic-krch) [inline drivers/pwm/pwm_nrf_sw.c]: use `LOG_MODULE_REGISTER(pwm_nrf_sw, CONFIG_PWM_LOG_LEVEL);`

File: drivers/pwm/pwm_nrf_sw.c [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+#define PWM_MAP_SIZE  DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT
+
+/* Nordic TIMER peripherals allow prescalers 0-9. */
+#define MAX_TIMER_PRESCALER  9
+/* Nordic TIMERs can be 16- or 32-bit wide. */
+#define MAX_TIMER_VALUE  (GET_TIMER_CAPABILITY(MAX_SIZE) == 32 ? UINT32_MAX \
+							       : UINT16_MAX)
+
+#if (GET_TIMER_CAPABILITY(CC_NUM) == 6)
+#define PWM_PERIOD_TIMER_CHANNEL  5
```
Comment (nordic-krch) [inline drivers/pwm/pwm_nrf_sw.c]: as i don't like ifdefs i would do `#define PWM_PERIOD_TIMER_CHANNEL ((GET_TIMER_CAPABILITY(CC_NUM) == 6) ? 5 : 3)`

**Comment by anangl** [comment]
Rebased on master.

File: drivers/pwm/pwm_nrf_sw.c (hunk 6 above)
Comment (anangl) [reply drivers/pwm/pwm_nrf_sw.c]: done

File: drivers/pwm/pwm_nrf_sw.c (hunk 7 above)
Comment (anangl) [reply drivers/pwm/pwm_nrf_sw.c]: done

**Comment by nordic-krch** [comment]
@anangl is there any way to detect if there is conflict in TIMER instance usage?

**Comment by anangl** [comment]
Replaced (as a fix for #15323) by #20369.
